helper/schema/resource_diff.go

```diff
@@ -0,0 +1,453 @@
+// any possibility of conflicts, but can be called on its own to just remove a
+// specific key from the diff completely.
+//
+// Note that this does not wipe an override. This function is only allowed on
+// computed keys.
+func (d *ResourceDiff) Clear(key string) error {
+	if !d.schema[key].Computed {
+		return fmt.Errorf("Clear is allowed on computed attributes only - %s is not one", key)
+	}
+
+	return d.clear(key)
+}
+
+func (d *ResourceDiff) clear(key string) error {
+	// Check the schema to make sure that this key exists first.
+	if _, ok := d.schema[key]; !ok {
+		return fmt.Errorf("%s is not a valid key", key)
+	}
+	for k := range d.diff.Attributes {
+		if strings.HasPrefix(k, key) {
+			delete(d.diff.Attributes, k)
+		}
+	}
+	return nil
+}
+
+// diffChange helps to implement resourceDiffer and derives its change values
+// from ResourceDiff's own change data, in addition to existing diff, config, and state.
+func (d *ResourceDiff) diffChange(key string) (interface{}, interface{}, bool, bool) {
+	old, new := d.getChange(key)
+	// log.Printf("\nkey:%s\n\nold:%s\n\nnew:%s\n", key, spew.Sdump(old), spew.Sdump(new))
+
+	if !old.Exists {
+		old.Value = nil
+	}
+	if !new.Exists {
+		new.Value = nil
+	}
+
+	return old.Value, new.Value, !reflect.DeepEqual(old.Value, new.Value), new.Computed
+}
+
+// SetNew is used to set a new diff value for the mentioned key. The value must
+// be correct for the attribute's schema (mostly relevant for maps, lists, and
+// sets). The original value from the state is used as the old value.
+//
+// This function is only allowed on computed attributes.
+//
+// It is an unsupported operation to set invalid values with this function -
+// doing so will taint any existing diff for this key and will remove it from
+// the catalog.
+func (d *ResourceDiff) SetNew(key string, value interface{}) error {
+	return d.SetDiff(key, d.get(strings.Split(key, "."), "state").Value, value, false)
+}
+
+// SetNewComputed functions like SetNew, except that it sets the new value to
+// the zero value and flags the attribute's diff as computed.
+//
+// This function is only allowed on computed keys.
+func (d *ResourceDiff) SetNewComputed(key string) error {
+	return d.SetDiff(key, d.get(strings.Split(key, "."), "state").Value, d.schema[key].ZeroValue(), true)
+}
+
+// SetDiff allows the setting of both old and new values for the diff
+// referenced by a given key. This can be used to completely override
+// Terraform's own diff behaviour, and can be used in conjunction with Clear or
+// ClearAll to construct a compleletely new diff based off of provider logic
+// alone.
+//
+// This function is only allowed on computed keys.
+func (d *ResourceDiff) SetDiff(key string, old, new interface{}, computed bool) error {
+	if !d.schema[key].Computed {
+		return fmt.Errorf("SetNew, SetNewComputed, and SetDiff are allowed on computed attributes only - %s is not one", key)
+	}
+
+	return d.setDiff(key, old, new, computed)
```

Comment:
I'm fine with that :) it makes sense to completely offload old value calculation to refresh/read for sure. If we do this too, we can get rid of the old value writer layer as well and just use state for old values.